File: backend/src/main/java/com/dongnemarket/product/event/ProductFavoriteCountHandler.java
package com.dongnemarket.product.event;

import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import com.dongnemarket.global.common.event.FavoriteAddedEvent;
import com.dongnemarket.global.common.event.FavoriteRemovedEvent;
import com.dongnemarket.product.repository.ProductRepository;

@Component
public class ProductFavoriteCountHandler {

	private final ProductRepository productRepository;

	public ProductFavoriteCountHandler(ProductRepository productRepository) {
		this.productRepository = productRepository;
	}

	@EventListener
	public void on(FavoriteAddedEvent event) {
		productRepository.incrementFavoriteCount(event.getProductId());
	}

	@EventListener
	public void on(FavoriteRemovedEvent event) {
		productRepository.decrementFavoriteCount(event.getProductId());
	}
}

File: backend/src/main/kotlin/com/dongnemarket/product/event/ProductFavoriteCountHandler.kt
package com.dongnemarket.product.event

import com.dongnemarket.global.common.event.FavoriteAddedEvent
import com.dongnemarket.global.common.event.FavoriteRemovedEvent
import com.dongnemarket.product.repository.ProductRepository
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

/** 찜 추가·삭제 이벤트를 받아 `Product.favoriteCount` 를 벌크 update 로 증감시킨다. */
@Component
class ProductFavoriteCountHandler(
    private val productRepository: ProductRepository,
) {
    @EventListener
    fun on(event: FavoriteAddedEvent) {
        productRepository.incrementFavoriteCount(event.productId)
    }

    @EventListener
    fun on(event: FavoriteRemovedEvent) {
        productRepository.decrementFavoriteCount(event.productId)
    }
}
